internal/controller/ngrok/cloudendpoint_controller.go

```diff
@@ -0,0 +1,342 @@
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+package ngrok
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/go-logr/logr"
+	"github.com/ngrok/ngrok-api-go/v6"
+	ingressv1alpha1 "github.com/ngrok/ngrok-operator/api/ingress/v1alpha1"
+	ngrokv1alpha1 "github.com/ngrok/ngrok-operator/api/ngrok/v1alpha1"
+	"github.com/ngrok/ngrok-operator/internal/controller"
+	"github.com/ngrok/ngrok-operator/internal/ngrokapi"
+)
+
+const (
+	trafficPolicyNameIndex = "spec.trafficPolicyName"
+)
+
+// CloudEndpointReconciler reconciles a CloudEndpoint object
+type CloudEndpointReconciler struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	controller *controller.BaseController[*ngrokv1alpha1.CloudEndpoint]
+
+	Log            logr.Logger
+	Recorder       record.EventRecorder
+	NgrokClientset ngrokapi.Clientset
+}
+
+// SetupWithManager sets up the controller with the Manager.
+// It also sets up a Field Indexer to index Cloud Endpoints by their Traffic Policy name
+// Additionally, this triggers updates when a trafficPolicy is created or updated but not when deleted
+func (r *CloudEndpointReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if r.NgrokClientset == nil {
+		return errors.New("NgrokClientset is required")
+	}
+
+	r.controller = &controller.BaseController[*ngrokv1alpha1.CloudEndpoint]{
+		Kube:     r.Client,
+		Log:      r.Log,
+		Recorder: r.Recorder,
+
+		StatusID: func(clep *ngrokv1alpha1.CloudEndpoint) string { return clep.Status.ID },
+		Create:   r.create,
+		Update:   r.update,
+		Delete:   r.delete,
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &ngrokv1alpha1.CloudEndpoint{}, trafficPolicyNameIndex, func(o client.Object) []string {
+		clep, ok := o.(*ngrokv1alpha1.CloudEndpoint)
+		if !ok {
+			return nil
+		}
+		return []string{clep.Spec.TrafficPolicyName}
+	}); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ngrokv1alpha1.CloudEndpoint{}).
+		Watches(
+			&ngrokv1alpha1.NgrokTrafficPolicy{},
+			handler.EnqueueRequestsFromMapFunc(r.findCloudEndpointForTrafficPolicy),
```

Comment:
We have a new function to help create/manage these from the base controller that also helps retain the log ctx: https://github.com/ngrok/ngrok-operator/blob/main/internal/controller/base_controller.go#L122

Would you mind trying out this function?